Tweak teleport staff

D-Cysteine opened the pull request:
New controls for the teleport staff: R-click will try to teleport to an anchor (the current shift-R-click behavior), and teleport to look (the current R-click behavior) if no anchor was found. Shift-R-click will teleport to look, ignoring anchors. Additionally, anchors will now be hidden while sneaking, so that they don't obstruct your view.

I don't know how people use the staff, so please do complain here if these changes would make it harder to use. There is also a config option to restore the original controls.

Also expand the teleport through wall range from 3 to 4, to help with being able to double-click on a distant wall to teleport to, then through it quickly.

AbdielKavash commented:
> I don't know how people use the staff, so please do complain here if these changes would make it harder to use.

> R-click will try to teleport to an anchor (the current shift-R-click behavior), and teleport to look (the current R-click behavior) if no anchor was found.

This sounds like it would end up with accidental teleports to unwanted places if you just miss the anchor.

> Additionally, anchors will now be hidden while sneaking, so that they don't obstruct your view.

This sounds backwards to me. I only want to see anchors when I actually want to teleport; a majority of time I would like them to be hidden, so that *they* do not obstruct my view.

D-Cysteine commented:
> > I don't know how people use the staff, so please do complain here if these changes would make it harder to use.
> 
> > R-click will try to teleport to an anchor (the current shift-R-click behavior), and teleport to look (the current R-click behavior) if no anchor was found.
> 
> This sounds like it would end up with accidental teleports to unwanted places if you just miss the anchor.

Yup, this is (IMHO) the potential problem with this change. I felt that the ease-of-use from being able to easily do both types of teleports without needing to hold shift offsets this, but if it doesn't, then this would be a bad change. That said, if you miss, you do end up teleporting where you are looking, which should in most cases be towards the anchor anyway, so you can just click again.

> 
> > Additionally, anchors will now be hidden while sneaking, so that they don't obstruct your view.
> 
> This sounds backwards to me. I only want to see anchors when I actually want to teleport; a majority of time I would like them to stay hidden, so that they do not obstruct my view while working.

The current behavior is to always show anchors while holding the staff (active hotbar item), even though you can only teleport to them while holding shift. The new behavior will hide the anchors while holding shift (and holding the staff), because you can't teleport to them while holding shift. So we show the anchors less often with the new change.

> 
> > There is also a config option to restore the original controls.
> 
> This is maybe risking bloat a bit, but how about making two configs: one for right-click, and one for sneak-right-click; and allow the player to bind both to: "teleport to anchor", "teleport forward", "teleport to anchor if available, otherwise forward", or "none"?

Good idea; added. I ended up keeping the original config too to make it easy to switch back, and also because having it on slightly changes how travel anchors are shown (they are always shown, regardless of holding shift).

---

As I was testing this, I kept finding more and more bugs, which I tried to fix, but this PR ended up growing a lot. I'll update the PR description with all of the stuff I changed.

AbdielKavash commented on src/main/java/crazypants/enderio/teleport/anchor/TileTravelAnchor.java:
```diff
@@ -346,10 +346,4 @@ public void invalidate() {
-    public void onChunkUnload() {
-        super.onChunkUnload();
-        TravelController.instance.travelDestinations.remove(worldObj.provider.dimensionId, getLocation());
-    }
```
Since you are no longer removing the anchor from `travelDestinations` on chunk unload, please make sure that repeatedly loading and unloading a chunk with an anchor doesn't continuously place more copies of the same anchor into this multimap. I am not well-versed enough in vanilla code to understand when exactly `validate` and `invalidate` are called, sorry.

D-Cysteine replied:
I think I checked that `validate()` doesn't get called multiple times; I think it's just when the block gets placed (and conversely, `invalidate()` when the block gets broken). IIRC that was the original source of the bug: if a travel anchor gets unloaded, the original code would remove it from `travelDestinations`, but never re-add it when it gets loaded again.

That said, `travelDestinations` is a `SetMultimap`, which does not allow duplicate entries, so adding the same `BlockCoord` to it multiple times will not result in duplicates anyway. I've added a comment explaining this in `validate()`.

AbdielKavash replied:
My bad, I misread the structure as just a `MultiMap`. 👍

AbdielKavash reviewed (changes requested):
Extra comment:

Even though you can now teleport to unloaded anchors, they are not visible when holding the staff. This means that (a) if you actually want to use this feature, you have to remember where your anchor is, and blindly aim in the right direction; and (b) you can accidentally teleport to a far away anchor that you did not see when trying to use the forward teleport.

Unloaded anchors should therefore be made visible (this might require a more significant rework, as the client doesn't have the TE anymore), or this feature reconsidered.

D-Cysteine commented on src/main/java/crazypants/enderio/teleport/packet/PacketLongDistanceTravelEvent.java:
```diff
@@ -69,51 +95,105 @@ public IMessage onMessage(PacketLongDistanceTravelEvent message, MessageContext
 
-    public static boolean doServerTeleport(Entity toTp, boolean conserveMotion, TravelSource source) {
-        EntityPlayer player = toTp instanceof EntityPlayer ? (EntityPlayer) toTp : null;
+    public static boolean doServerTeleport(Entity toTp, boolean conserveMotion, TravelSource source,
+            boolean alsoDoTeleport, int tpX, int tpY, int tpZ) {
+        if (!(toTp instanceof EntityPlayer)) {
+            return false;
+        }
+        EntityPlayer player = (EntityPlayer) toTp;
```
We actually validate on line 84 of this file that it's teleporting a player, so it is indeed impossible to use it to teleport non-players. This is also true of `PacketTravelEvent` (which this class was based off of).

You're right that it would probably be best to change the method signature, which would also allow removing the `entityId` field. I would prefer to do this in a separate PR, and refactor both packets at once, though, as this PR is probably too large and tangled already. Let me know if you really want me to do it in this PR, though (and if you want both packets refactored).

D-Cysteine commented on src/main/java/crazypants/enderio/config/Config.java:
```diff
@@ -1269,6 +1305,37 @@ public static void processConfig(Configuration config) {
                 .getInt(teleportStaffFailedBlinkDistance);
 
+        teleportStaffOriginalControls = config.get(
+                sectionPersonal.name,
```
Hmm, I was originally thinking that it should be placed in Personal Settings because it looked like Staff Settings was all server-side-affecting configs (e.g. things like max range and energy consumption), whereas this one was purely client-side. But it looks like there's some other client-side configs in there too, like render size, so I moved the new configs. Thanks!

D-Cysteine commented on src/main/java/crazypants/enderio/config/Config.java:
```diff
@@ -1269,6 +1305,37 @@ public static void processConfig(Configuration config) {
+                .getBoolean(teleportStaffOriginalControls);
+
+        teleportStaffAction = config.get(
+                sectionPersonal.name,
```
Moved

D-Cysteine commented on src/main/java/crazypants/enderio/config/Config.java:
```diff
@@ -174,6 +174,42 @@ public String lc() {
      */
     public static int teleportStaffFailedBlinkDistance = 64;
+    /** If true, use the original staff of teleportation controls. */
+    public static boolean teleportStaffOriginalControls = false;
```
There's actually a subtle difference, in that this setting enables the old render behavior (always show travel anchors), which is not settable using the new configs. That said, you're probably right that this minor difference doesn't warrant an entire separate config setting, so I've removed this.

D-Cysteine commented on src/main/java/crazypants/enderio/teleport/TravelController.java:
```diff
@@ -190,8 +193,27 @@ public void addBlockToBlinkBlackList(String blockName) {
+
+    public boolean activateTravelAccessable(ItemStack equipped, World world, EntityPlayer player, TravelSource source,
+            boolean alsoDoTeleport) {
         if (!hasTarget()) {
             PacketLongDistanceTravelEvent p = new PacketLongDistanceTravelEvent(player, false, source);
+            if (alsoDoTeleport) {
+                Optional<BlockCoord> destinationOptional = findTeleportDestination(player);
+                if (destinationOptional.isPresent()) {
+                    BlockCoord destination = destinationOptional.get();
+                    p = new PacketLongDistanceTravelEvent(
+                            player,
+                            false,
+                            source,
+                            true,
+                            destination.x,
+                            destination.y,
+                            destination.z);
+                }
+            }
```
I think that doesn't quite work, because we actually have two nested `if`-statements (lines 203 and 205). We'd have to either have duplicate `else`-clauses, or always call the expensive `findTeleportDestination()` method even if it's not needed.

I think the extra allocation is probably cheap enough not to warrant the duplicate `else`-clauses, so I've left this as-is. We could also avoid the extra allocation by using `Supplier<PacketLongDistanceTravelEvent>` instead, though I don't know if that would actually end up even more expensive due to allocating a `Supplier` and adding a call to `Supplier.get()`.

AbdielKavash replied:
```java
            PacketLongDistanceTravelEvent p;
            if (alsoDoTeleport) {
                Optional<BlockCoord> destinationOptional = findTeleportDestination(player);
                if (destinationOptional.isPresent()) {
                    BlockCoord destination = destinationOptional.get();
                    p = new PacketLongDistanceTravelEvent(
                            player,
                            false,
                            source,
                            true,
                            destination.x,
                            destination.y,
                            destination.z);
                }
            }
            if (p == null) {
                // Forward teleport disabled or impossible, only teleport to an anchor.
                p = new PacketLongDistanceTravelEvent(player, false, source);
            }
```

D-Cysteine replied:
Sorry, I missed your most recent comment. How's this?

AbdielKavash replied:
Oh, that's even simpler 👍

D-Cysteine commented on src/main/java/crazypants/enderio/teleport/TravelController.java:
```diff
@@ -371,49 +411,37 @@ public boolean doTeleport(EntityPlayer player) {
 
-    /** Returns empty optional if no travel destination was found. */
+    /** Finds a long-distance travel anchor, or returns empty optional if unsuccessful. */
     public Optional<BlockCoord> findTravelDestination(EntityPlayer player, TravelSource source) {
```
I prefer to return `Optional` rather than `null` from methods that may not return a value, in order to force explicit handling of the `null` case, as well as being completely clear about the return value possibly being `null`. Isn't that the main point of `Optional`?

Yeah, I think the methods are in a rather nonsensical order, since I think they've kind of gotten added over time 😓. I've moved this method (and `findTeleportDestination()`, right above it) as you requested.

AbdielKavash replied:
`Optional` generally makes sense when both `null` and "unset"/"unknown" are valid return values with different semantics. Here the result is either a valid `BlockCoord`, or "no valid destination found". There is no need for a different third state.

Are these methods ever used (or meant to be used) outside of `TravelController`? Maybe they should be made private so that you don't need to worry about other code using it incorrectly.

AbdielKavash replied:
Fair enough, your reasons make sense here.

D-Cysteine commented on src/main/java/crazypants/enderio/teleport/TravelController.java:
```diff
@@ -619,6 +624,16 @@ public boolean isTravelItemActive(EntityPlayer ep, boolean checkInventoryAndBaub
+        if (getTravelItemTravelSource(ep, false) != TravelSource.TELEPORT_STAFF
+                || Config.teleportStaffOriginalControls) {
+            return false;
+        }
```
This hiding feature is new to this PR, and I didn't want to change the behavior of anything other than the staff of teleportation, so it is intended that we do not handle any other travel items here.

I've added a comment explaining this.

AbdielKavash replied:
I would probably still prefer this to apply to other items too; by playing around with it for a while I liked the feature. But I'll let it stand; I'd encourage you to make a new PR for this too if you want to keep this one focused.

D-Cysteine commented on src/main/java/crazypants/enderio/teleport/TravelController.java:
```diff
@@ -914,55 +984,67 @@ private BlockCoord getValidVerticalTarget(EntityPlayer player, BlockCoord curren
-        if (candidates.isEmpty()) {
-            return;
-        }
+    /**
+     * Returns a row-sorted table with angle in radians as the row, distance as the column, and block coordinate as the
+     * value, holding all blocks which lie within the specified viewing angle and distance.
+     *
+     * <p>
+     * The table is sorted by angle ascending, so the smallest angle will be first.
+     *
+     * @param player      The player to search from.
+     * @param blocks      The blocks to search through.
+     * @param maxAngle    The maximum allowed angle, in radians.
+     * @param maxDistance The maximum allowed distance.
+     */
+    private RowSortedTable<Double, Double, BlockCoord> findBlocksWithinAngle(EntityPlayer player,
+            Iterable<BlockCoord> blocks, double maxAngle, double maxDistance) {
```
Yes:  `findTravelDestinations()`, on line 713 (well, before the re-arranging of functions) calls this, and uses more than just the first entry. The reason is that we want to show a failure message to the player if there is an anchor that is just barely out of range, but not if there's another anchor that maybe has a wider angle, but is actually within range.

I suppose we could change that behavior and avoid needing this, but IIRC that was from before this PR, and I hadn't inteded to change how that part worked in this PR, so I had to add the table.

AbdielKavash replied:
The default range, even for the simple Staff of Travel or other travel tools, is 128 blocks in GTNH. If an anchor is in the "barely outside" range, it is very likely to be just completely outside of the player's range and invisible anyway. It took me some effort to even trigger this message at all.

I'll leave it up to you, keep it as it is, remove it, or keep it for now and clean up in a later PR.

D-Cysteine replied:
It's mainly meant for the 2048-range staff of teleportation, where the extra 25% message range translates to 512 blocks.
Let's leave it for now; we can always remove it in a later PR.

D-Cysteine commented:
> Extra comment:
> 
> Even though you can now teleport to unloaded anchors, they are not visible when holding the staff. This means that (a) if you actually want to use this feature, you have to remember where your anchor is, and blindly aim in the right direction; and (b) you can accidentally teleport to a far away anchor that you did not see when trying to use the forward teleport.
> 
> Unloaded anchors should therefore be made visible (this might require a more significant rework, as the client doesn't have the TE anymore), or this feature reconsidered.

Yes, it would be nice if we could render the anchors at long distance, but I couldn't get that to work; IIRC there is some hard-coded (or otherwise difficult to bypass) limit on Minecraft render range that prevents this from working. It probably wouldn't work well to render anchors at 2048 range anyway; I think they'd either be too small to click on, or large enough to obstruct vision.

The intended way to work around this is to mark travel anchors with JourneyMap waypoints. Also note that this behavior predates this PR; the staff of teleportation has always worked this way.

Point (b) is a good point; I feel that the convenience of being able to do both actions in one click outweighs it though, and it's also fairly unlikely that there would be a mystery travel anchor that close to a player's base. If it becomes a problem, it's always possible to change the action configs to only teleport to look, and ignore anchors.

AbdielKavash commented:
> Point (b) is a good point; I feel that the convenience of being able to do both actions in one click outweighs it though, and it's also fairly unlikely that there would be a mystery travel anchor that close to a player's base. If it becomes a problem, it's always possible to change the action configs to only teleport to look, and ignore anchors.

The setting I was imagining is a multiplayer server, where you try to teleport around in world, and the 10° cone eventually covers a huge area in the distance, and you accidentally teleport to someone else's base. But the 2048 blocks max range should to a degree prevent this too.

Let's leave it as it is/for future iteration. I would still like the renderer to be tweaked to show distant anchors; though if this is complicated then maybe for another time.

AbdielKavash commented:
All looks good now. Nice work!
